packages/protocol/contracts/L1/libs/LibProposing.sol

```diff
@@ -71,28 +99,57 @@ library LibProposing {
         meta.depositsProcessed =
             LibEthDepositing.processDeposits(state, config, input.beneficiary);
 
-        unchecked {
-            meta.timestamp = uint64(block.timestamp);
-            meta.l1Height = uint64(block.number - 1);
-            meta.l1Hash = blockhash(block.number - 1);
-            meta.mixHash = bytes32(block.difficulty * state.numBlocks);
-        }
-
+        // Init the block
         TaikoData.Block storage blk =
-            state.blocks[state.numBlocks % config.ringBufferSize];
+            state.blocks[state.numBlocks % config.blockRingBufferSize];
 
+        blk.metaHash = LibUtils.hashMetadata(meta);
         blk.blockId = state.numBlocks;
-        blk.proposedAt = meta.timestamp;
+        blk.gasLimit = meta.gasLimit;
         blk.nextForkChoiceId = 1;
         blk.verifiedForkChoiceId = 0;
-        blk.metaHash = LibUtils.hashMetadata(meta);
+        blk.proverReleased = false;
+
         blk.proposer = msg.sender;
+        blk.feePerGas = state.feePerGas;
+        blk.proposedAt = meta.timestamp;
 
-        uint64 blockFee;
+        if (assignedProver == address(0)) {
+            if (state.numOpenBlocks >= config.rewardOpenMaxCount) {
+                revert L1_TOO_MANY_OPEN_BLOCKS();
+            }
+            blk.rewardPerGas = state.feePerGas;
+            ++state.numOpenBlocks;
+        } else {
+            blk.assignedProver = assignedProver;
+
+            // Cap the reward to a range of [95%, 105%] * blk.feePerGas, if
+            // rewardPerGasRange is set to 5% (500 bp)
+            uint32 diff = blk.feePerGas * config.rewardPerGasRange / 10_000;
+            blk.rewardPerGas = uint32(
+                uint256(rewardPerGas).min(state.feePerGas + diff).max(
+                    state.feePerGas - diff
+                )
+            );
+
+            blk.proofWindow = uint16(
+                uint256(state.avgProofDelay * 3).min(config.proofMaxWindow).max(
```

Comment:
*3 maybe too long. The prover with a high staking amount may use a less efficient machine to reduce cost, and the average proof delay may increase. This design has no encouragement to prover of faster speed. There may be some concern that all provers use cheap machines and have long proof time. That means if the provers found that the total cost of more but less efficient machines is less than less but highly efficient machines, the provers will choose the former, and the average proof time won't be reduced gradually. Though we are not clear about this yet, and it requires more investigation about the price with different machines.

Reply:
Right, we are not encouraging fast proofs now, which is a pending issue. Will work on it later. This is also mentioned in https://github.com/taikoxyz/taiko-mono/issues/14090.